### PR TITLE
Fix bug

### PR DESCRIPTION
**Pull request by VladDalidovich**
Change BLL

### CODE REVIEW

File: Godel.CM.BLL/Services/CurrencyService.cs [hunk 1]
```diff
@@ -42,21 +42,44 @@ public async Task ExchangeRateUploadAsync(IEnumerable<RequestExchangeRateModel>
                 throw new ArgumentNullException(nameof(exchangeRates));
             }
 
-            var convertRates = exchangeRates.Select(ExchangeRateConverter.ToExchangeRate);
+            List<Currency> necessaryCurrency = new List<Currency>();
+            foreach (var el in exchangeRates)
+            {
+                necessaryCurrency.Add(await currencyRepository.GetSingleAsync(c => c.Code == el.Code));
```
Comment (HenadziBaranau) [inline Godel.CM.BLL/Services/CurrencyService.cs]: Don't do requests to database for each code. Pass range of codes to query all currencies by code at once

File: Godel.CM.BLL/Services/CurrencyService.cs [hunk 2]
```diff
@@ -42,21 +42,44 @@ public async Task ExchangeRateUploadAsync(IEnumerable<RequestExchangeRateModel>
...
+            }
+
+            List<ExchangeRate> convertRates = new List<ExchangeRate>();
+            foreach (var el in exchangeRates)
+            {
+                convertRates.Add(el.ToExchangeRate(necessaryCurrency.Find(c => c.Code == el.Code).Id));
```
Comment (HenadziBaranau) [inline Godel.CM.BLL/Services/CurrencyService.cs]: Just think about how to debug this.

File: Godel.CM.BLL/Services/CurrencyService.cs [hunk 3]
```diff
@@ -42,21 +42,44 @@ public async Task ExchangeRateUploadAsync(IEnumerable<RequestExchangeRateModel>
...
+            return result.ToResponseCurrencyModel(lastRate.ToResponseExchangeRateModel());
         }
 
         public async Task<IEnumerable<ResponseCurrencyModel>> GetCurrencyListAsync()
         {
             var result = await currencyRepository.GetAllAsync();
             return result.Select(CurrencyConverter.ToResponseCurrencyModel);
         }
+
+        private async Task<ExchangeRate> GetLastRate(int currencyId, DateTime date)
```
Comment (HenadziBaranau) [inline Godel.CM.BLL/Services/CurrencyService.cs]: Should be async

File: Godel.CM.BLL/Services/CurrencyService.cs [hunk 4]
```diff
@@ -42,21 +42,44 @@ public async Task ExchangeRateUploadAsync(IEnumerable<RequestExchangeRateModel>
...
+
             exchangeRateRepository.AddRange(convertRates);
             await unitOfWork.SaveChangesAsync();
         }
 
         public async Task<ResponseCurrencyModel> GetCurrencyAsync(string currencyCode)
         {
             var result = await currencyRepository.GetSingleAsync(c => c.Code == currencyCode);
-            return result.ToResponseCurrencyModel();
+            var lastRate = await GetLastRate(result.Id, Convert.ToDateTime(DateTime.Now.Date.ToString("dd/MM/yyyy")));
```
Comment (HenadziBaranau) [inline Godel.CM.BLL/Services/CurrencyService.cs]: DateTime.Now should be moved to IDateTimeProvider. Ask @Werocky1502 about it.

File: Godel.CM.BLL/Services/CurrencyService.cs [hunk 5]
```diff
@@ -42,21 +42,44 @@ public async Task ExchangeRateUploadAsync(IEnumerable<RequestExchangeRateModel>
...
             var result = await currencyRepository.GetAllAsync();
             return result.Select(CurrencyConverter.ToResponseCurrencyModel);
         }
+
+        private async Task<ExchangeRate> GetLastRate(int currencyId, DateTime date)
+        {
+            var lastRate = await exchangeRateRepository.GetSingleOrDefaultAsync(r => r.CurrencyId == currencyId && r.UpdateDate == date);
+            if (lastRate == null)
+            {
+                return await GetLastRate(currencyId, date.AddMonths(-1));
```
Comment (HenadziBaranau) [inline Godel.CM.BLL/Services/CurrencyService.cs]: -1 - magic number. Just query really last date

File: Godel.CM.BLL/Services/CurrencyService.cs [hunk 6]
```diff
@@ -42,21 +42,44 @@ public async Task ExchangeRateUploadAsync(IEnumerable<RequestExchangeRateModel>
                 throw new ArgumentNullException(nameof(exchangeRates));
             }
 
-            var convertRates = exchangeRates.Select(ExchangeRateConverter.ToExchangeRate);
+            List<Currency> necessaryCurrency = new List<Currency>();
```
Comment (HannaKeryna) [inline Godel.CM.BLL/Services/CurrencyService.cs]: You can use var 
Please read code style conventions
https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/inside-a-program/coding-conventions

File: Godel.CM.Data.Mapper/CurrencyConverter.cs [hunk 7]
```diff
@@ -1,17 +1,28 @@
...
 using Godel.CM.Data.ViewModels.Request;
 using Godel.CM.Data.ViewModels.Response;
+using System.Collections.Generic;
 
 namespace Godel.CM.Data.Mapper
 {
     public static class CurrencyConverter
     {
-        public static ResponseCurrencyModel ToResponseCurrencyModel(this Currency model)
+        public static ResponseCurrencyModel ToResponseCurrencyModel(this Currency model, ResponseExchangeRateModel rate)
```
Comment (HenadziBaranau) [inline Godel.CM.Data.Mapper/CurrencyConverter.cs]: Pass domain rate here, and do `ToResponseExchangeRateModel`

File: Godel.CM.BLL/Services/CurrencyService.cs [hunk 8]
```diff
@@ -42,21 +42,44 @@ public async Task ExchangeRateUploadAsync(IEnumerable<RequestExchangeRateModel>
...
+            foreach (var el in exchangeRates)
+            {
+                necessaryCurrency.Add(await currencyRepository.GetSingleAsync(c => c.Code == el.Code));
+            }
+
+            List<ExchangeRate> convertRates = new List<ExchangeRate>();
+            foreach (var el in exchangeRates)
```
Comment (HannaKeryna) [inline Godel.CM.BLL/Services/CurrencyService.cs]: You can convert foreach to LINQ

File: Godel.CM.BLL/Services/CurrencyService.cs [hunk 9]
```diff
@@ -42,21 +42,44 @@ public async Task ExchangeRateUploadAsync(IEnumerable<RequestExchangeRateModel>
...
+
+        private async Task<ExchangeRate> GetLastRate(int currencyId, DateTime date)
+        {
+            var lastRate = await exchangeRateRepository.GetSingleOrDefaultAsync(r => r.CurrencyId == currencyId && r.UpdateDate == date);
+            if (lastRate == null)
+            {
+                return await GetLastRate(currencyId, date.AddMonths(-1));
+            }
+
+            return lastRate;
```
Comment (HannaKeryna) [inline Godel.CM.BLL/Services/CurrencyService.cs]: You can use ?? and remove if:
    return lastRate ?? await GetLastRate(currencyId, date.AddMonths(-1));

**Pull request merged**